engine/classes/Elgg/Values.php

```diff
@@ -46,10 +47,106 @@ public static function getArray() {
 		return [];
 	}
 
+	/**
+	 * Returns timestamp value of the time representation
+	 *
+	 * @param DateTime|string|int $time Time
+	 *
+	 * @return int
+	 * @throws \DataFormatException
+	 */
+	public static function normalizeTimestamp($time) {
+		try {
+			if ($time instanceof \DateTime) {
+				return $time->getTimestamp();
+			} else if (is_string($time)) {
+				$dt = new \DateTime($time);
+
+				return $dt->getTimestamp();
+			}
+		} catch (\Exception $e) {
+			throw new \DataFormatException($e->getMessage());
+		}
+
+		return (int) $time;
+	}
+
+	/**
+	 * Prepare IDs
+	 *
+	 * @param array ...$args IDs
+	 *
+	 * @return int[]
+	 * @throws \DataFormatException
+	 */
+	public static function normalizeIds(...$args) {
+		if (empty($args)) {
+			return ELGG_ENTITIES_ANY_VALUE;
+		}
+
+		$ids = [];
+		foreach ($args as $arg) {
+			if (!isset($arg)) {
+				continue;
+			}
+			if (is_object($arg) && isset($arg->id)) {
+				$ids[] = (int) $arg->id;
+			} else if (is_array($arg)) {
+				foreach ($arg as $a) {
+					$el_ids = self::normalizeIds($a);
+					$ids = array_merge($ids, $el_ids);
+				}
+			} else if (is_numeric($arg)) {
+				$ids[] = (int) $arg;
+			} else {
+				$arg = print_r($arg, true);
+				throw new \DataFormatException("Parameter '$arg' can not be resolved to a valid ID'");
+			}
+		}
+
+		return array_unique($ids);
+	}
+
+	/**
+	 * Flatten an array of data into an array of GUIDs
+	 *
+	 * @param mixed ...$args Elements to normalize
+	 *
+	 * @return int[]|null
+	 * @throws \DataFormatException
+	 */
+	public static function normalizeGuids(...$args) {
```

Comment:
what are the `...` doing here?

Reply:
http://php.net/manual/en/functions.arguments.php#functions.variable-arg-list

Reply:
thanks